### PR TITLE
[WIP] cupy.sparse

### LINKED ISSUE #1260: Support sparse matrix
Especially in NLP, we use sparse information such as in `EmbedID`. Its current implementation uses dense matrix to store gradient. When minibatch size is small, almost all elements of gradient are zero. Sparse matrices are more efficient in such situation.
- [ ] Use sparse matrices such as scikit-sparse in Chainer
- [x] Make sparse matrices with CUDA using cuSPARSE

### PR DESCRIPTION
**Pull request by unnonouno**
I'm implementing sparse matrix library which is compatible with scipy.sparse using cuSPARSE.
fix #1260

### CODE REVIEW

File: cupy/cuda/cupy_cusparse.h [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#ifndef INCLUDE_GUARD_CUPY_CUSPARSE_H
+#define INCLUDE_GUARD_CUPY_CUSPARSE_H
+
+#include <cusparse.h>
+
+#endif  // INCLUDE_GUARD_CUPY_CUSPARSE_H
```
Comment (gwtnb) [inline cupy/cuda/cupy_cusparse.h]: This file should be merged with `cupy_cuda.h`. Why there is `cupy_cudnn.h` that cuDNN is not distributed in CUDA itself.

Comment (unnonouno) [inline cupy/cuda/cupy_cusparse.h]: Oh, i see.

**Comment by cdicle** [comment]
Hi @unnonouno, 

I have a quick question. I have seen you implemented the sparse matrices in chainer if I am not mistaken. Can I define sparse convolution filters using your fork? 

I have never used chainer before, so excuse my ignorance. I am looking for a framework/implementation where I can individually pick the locations of nonzero entries in  convolution filters. The rest will stay zero through out. 

Looking forward to hearing from you,
Best,
Cha.

**Comment by unnonouno** [comment]
Oh, sorry for late reply. Currently I have no concrete idea to integrate `cupy.sparse` with Chainer, though we can support sparse matrixes in Chainer with more development.
This patch only provides a sparse matrix library which is compatible with `scipy.sparse`. We need to integrate it to Chainer to support back-propagation.

**Comment by unnonouno** [comment]
I'll submit cusparse related PRs to cupy/cupy: https://github.com/cupy/cupy/issues/36

**Pull request closed without merging**